drivers/bus/acpi/compbatt/compbatt.c

```diff
@@ -202,40 +526,665 @@ CompBattDisableStatusNotify(IN PCOMPBATT_DEVICE_EXTENSION DeviceExtension)
+
+        return STATUS_NO_SUCH_DEVICE;
+    }
+
+    /*
+     * Before we are setting up new status wait notification points we need to
+     * refresh the composite status so that we get to know what values should be
+     * set for the current notification wait status.
+     */
+    Status = CompBattQueryStatus(DeviceExtension,
+                                 BatteryTag,
+                                 &BatteryStatus);
+    if (!NT_SUCCESS(Status))
+    {
+        if (CompBattDebug & COMPBATT_DEBUG_ERR)
+        {
+            DbgPrint("CompBatt: Failed to refresh composite battery's status (Status 0x%08lx)\n", Status);
+        }
+
+        return Status;
+    }
+
+    /* Print out battery status data that has been polled */
+    if (CompBattDebug & COMPBATT_DEBUG_INFO)
+    {
+        DbgPrint("CompBatt: Latest composite battery status (when setting notify status)\n"
+                 "          PowerState -> 0x%lx\n"
+                 "          Capacity -> %u\n"
+                 "          Voltage -> %u\n"
+                 "          Rate -> %u\n",
+                 BatteryStatus.PowerState,
+                 BatteryStatus.Capacity,
+                 BatteryStatus.Voltage,
+                 BatteryStatus.Rate);
+    }
+
+    /* Calculate the high and low capacity differences based on the real summed capacity of the composite */
+    LowCapDifference = DeviceExtension->BatteryStatus.Capacity - BatteryNotify->LowCapacity;
+    HighCapDifference = BatteryNotify->HighCapacity - DeviceExtension->BatteryStatus.Capacity;
+
+    /* Cache the notification parameters provided for later usage when polling for battery status */
+    DeviceExtension->WaitNotifyStatus.PowerState = BatteryNotify->PowerState;
+    DeviceExtension->WaitNotifyStatus.LowCapacity = BatteryNotify->LowCapacity;
+    DeviceExtension->WaitNotifyStatus.HighCapacity = BatteryNotify->HighCapacity;
+
+    /* Toggle the valid notify flag as these are the newer notification settings */
+    DeviceExtension->Flags |= COMPBATT_STATUS_NOTIFY_SET;
+
+    /*
+     * Get the number of currently linked batteries to composite and total rate,
+     * we will use these counters later to determine the wait values for each
+     * individual battery.
+     */
+    BadBattery = CompBattCalculateTotalRateAndLinkedBatteries(DeviceExtension,
+                                                              &TotalRate,
+                                                              &BatteriesCount);
+
+    /*
+     * Of course we have to be sure that we have at least one battery linked
+     * with the composite battery at this time of getting invoked to set new
+     * notification wait settings.
+     */
+    ASSERT(BatteriesCount != 0);
+
+    /* Walk over the linked batteries list and set up new wait configuration settings */
+    ExAcquireFastMutex(&DeviceExtension->Lock);
+    ListHead = &DeviceExtension->BatteryList;
+    NextEntry = ListHead->Flink;
+    while (NextEntry != ListHead)
```

Comment:
See suggestion up above.
+ same suggestion for the 2 - 3 other similar loops in the code below.